1070 allow selection of download file output format

juzen2003 opened the pull request:
- Fixes #1070 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - If YES:
    - Database used: re-import with the latest pds-webtools
    - All Django tests pass: Y
    - FLAKE8 run on modified code: Y
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y (cart.js)
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
- Add a select input (default: zip) in cart download pane for users to select download file format (zip/tar/tgz).
- Add the same zip file tests for the other two formats. 

Known problems:

juzen2003 committed:
Allow users to select the format (zip/tar/tgz) of cart download files.
- In settings.py (line 364-371), create DOWNLOAD_FORMAT to store all the
  supported download files format.
- In cart.html (line 44-51), create the options for select tag based on
  DOWNLOAD_FORMAT
- In urls.py (line 19-20), update the api/download/opus_id.fmt url to
  support zip/tar/tgz.
- In cart.js (line 352-354), get the download files format from select
  input, and pass it in the api url.
- In views.py (line 596-752), in api_create_downloa, rename all 'zip'
  related variables to 'compressed' related. Also update the logic to
  instantiate & write the compressed file instance with zipfile or
  tarfile based on the file format.

juzen2003 committed:
Update _run_zip_file_equal and change it to _run_cmp_file_equal to work
with tar and tgz format.
- In api_test_helper.py (line 173-216), change _run_zip_file_equal to
  _run_cmp_file_equal, and make it work with tar and tgz as well.
- In settings.py (line 365-371), update DOWNLOAD_FORMAT to include both
  read and write modes for different file formats.
- In test_cart_api.py, apply the same tests for zip to tar & tgz.
- In views.py (line 762-763), rename 'zip_url' to 'cmp_url'.

juzen2003 committed:
Rename 'DOWNLOAD_FORMAT' to 'DOWNLOAD_FORMATS'

rfrenchseti commented on opus/application/apps/cart/templates/cart/cart.html:
```diff
@@ -41,7 +41,14 @@ <h1 class="op-download-options-title">Download Options</h1>
                         <i class="fas fa-info-circle" title="If unchecked, download products will be added to the zip file using the same directory structure used by the PDS RMS archive. If checked, all download products will be placed in the root directory of the zip file, unless there is an ambiguity caused by multiple files that have the same base filename. In that case those files only will b ...
                     </div>
-                <div class="col col-lg">Flat zip file structure</div>
+                <div class="col col-lg">Flat compressed file structure</div>
```
Let's remove the word "compressed" since a tar file isn't compressed. Just "Flat file structure".

rfrenchseti replied:
Could replace "cmp" with "archive" also through everything.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -19,6 +19,7 @@
 import logging
 import os
 import time
+import tarfile
```
Keep in alphabetical order - move above "time"

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -507,13 +509,13 @@ def api_reset_session(request):
 @never_cache
-def api_create_download(request, opus_id=None):
+def api_create_download(request, opus_id=None, fmt=None):
     r"""Creates a zip file of all items in the cart or the given OPUS ID.
```
Change "Creates a zip file" to "Creates an archive file"

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -507,13 +509,13 @@ def api_reset_session(request):
-        or: [__]api/download/(?P<opus_id>[-\w]+).zip
+        or: [__]api/download/(?P<opus_id>[-\w]+).(?P<fmt>zip|tar|tgz)
     Arguments: types=<PRODUCT_TYPES>
                urlonly=1 (optional) means to not zip the actual data products
```
"means to not include the actual..."   "means files in archive are stored..."

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -591,13 +593,18 @@ def api_create_download(request, opus_id=None):
+        fmt = request.GET.get('fmt', 'zip')
+
+    cmp_base_file_name = download_filename(opus_id, file_type) + f'.{fmt}'
+    cmp_root = cmp_base_file_name.split('.')[0]
```
Would be nicer to say:

cmp_root = download_filename(opus_id, file_type)
cmp_base_file_name = cmp_root + f'.{fmt'}

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -591,13 +593,18 @@ def api_create_download(request, opus_id=None):
+    cmp_base_file_name = download_filename(opus_id, file_type) + f'.{fmt}'
+    cmp_root = cmp_base_file_name.split('.')[0]
+    cmp_file_name = settings.TAR_FILE_PATH + cmp_base_file_name
+    # chksum_file_name = settings.TAR_FILE_PATH + f'checksum_{cmp_root}.txt'
```
We can get rid of this line now

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -591,13 +593,18 @@ def api_create_download(request, opus_id=None):
+    # If file format is not specified, set zip as the default format.
+    if not fmt:
+        fmt = request.GET.get('fmt', 'zip')
+
```
Since this is a public API, we should validate that fmt is a valid format and throw an HTTP404 if it isn't. Right now we will actually crash.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -632,12 +639,21 @@ def api_create_download(request, opus_id=None):
 
+    mime_type = settings.DOWNLOAD_FORMATS[fmt][0]
+    write_mode = settings.DOWNLOAD_FORMATS[fmt][1]
     # Add each file to the new zip file and create a manifest too
```
zip -> archive

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -632,12 +639,21 @@ def api_create_download(request, opus_id=None):
+        else:
+            cmp_file = tarfile.open(name=cmp_file_name, mode=write_mode)
+
     # chksum_fp = open(chksum_file_name, 'w')
```
Can go ahead and remove this line

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -721,22 +740,27 @@ def api_create_download(request, opus_id=None):
-    zip_file.write(csv_file_name, arcname='data.csv')
-    zip_file.write(url_file_name, arcname='urls.txt')
-    zip_file.close()
+    # cmp_file.write(chksum_file_name, arcname='checksum.txt')
```
Can remove this line

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -721,22 +740,27 @@ def api_create_download(request, opus_id=None):
+        cmp_file.add(url_file_name, arcname='urls.txt')
+    cmp_file.close()
 
     # os.remove(chksum_file_name)
```
Can remove this line

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -3,6 +3,7 @@
 import json
 import os
 import zipfile
+import tarfile
```
Keep in alphabetical order

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -169,28 +170,47 @@ def _run_csv_equal_file(self, url, exp_file):
         print(url)
         response = self._get_response(url)
         self.assertEqual(response.status_code, 200)
+        compressed_file_path = None
```
compressed -> archive throughout

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -2462,28 +2606,96 @@ def test__api_cart_download_multiple_no_duplicated_hierarchical(self):
+        expected = ['calibrated/COISS_2xxx/COISS_2002/data/1462783195_1462915477/N1462840881_1_CALIB.IMG', 'calibrated/COISS_2xxx/COISS_2002/data/1462783195_1462915477/N1462840881_1_CALIB.LBL', 'calibrated/COISS_2xxx/COISS_2008/data/1481264980_1481267140/N1481265970_1_CALIB.IMG', 'calibrated/COISS_2xxx/COISS_2008/data/1481264980_1481267140/N1481265970_1_CALIB.LBL', 'data.csv', 'manifest.csv', 'pr ...
+        self._run_cmp_file_equal(url, expected, fmt='tgz')
 
             #########################################################
```
Fix number of #

juzen2003 committed:
Update based on github feedback, rename all 'cmp_' related variables to
'archive_'. Add check/test for invalid fmt passing into __cart/download.json.
- In app_utils.py (line 331-335), add HTTP404_UNKNOWN_DOWNLOAD_FILE_FORMAT
  to return error message when an invalid fmt is passed into
  __cart/download.json.
- In test_cart_api.py (line 2644-2655), add test to check the error message
  when passing an invalid fmt to __cart/download.json.

juzen2003 committed:
Fixed the indentation at line 44 in cart.html.

juzen2003 committed:
Fixed the comments to get the symmetrical "#" at both ends at line 2658
in test_cart_api.py.
